File: backend/nestjs/clean-architecture/minimal/src/modules/audit-log/presentation/audit-logs.controller.ts
import { Controller, Get, Query, UseGuards } from '@nestjs/common';
import { QueryBus } from '@nestjs/cqrs';
import { ApiTags, ApiOperation, ApiResponse, ApiBearerAuth } from '@nestjs/swagger';
import { JwtAuthGuard } from '@/common/guards/jwt-auth.guard';
import { RolesGuard } from '@/common/guards/roles.guard';
import { Roles } from '@/common/decorators/roles.decorator';
import { PagedResult } from '@/common/types/paginated';
import { UserRole } from '@/modules/user-management/domain/value-objects/user-role';
import { GetAuditLogsQuery } from '../application/queries/get-audit-logs.query';
import { AuditLogDto, GetAuditLogsQueryDto } from '../application/dto/audit-log.dto';

@ApiTags('Audit Logs')
@ApiBearerAuth('JWT-auth')
@Controller('audit-logs')
@UseGuards(JwtAuthGuard, RolesGuard)
@Roles(UserRole.Admin)
export class AuditLogsController {
  constructor(private readonly queryBus: QueryBus) {}

  @Get()
  @ApiOperation({ summary: 'Get audit logs with pagination (Admin only)' })
  @ApiResponse({
    status: 200,
    description: 'Paginated audit logs retrieved successfully',
  })
  async getAuditLogs(@Query() queryDto: GetAuditLogsQueryDto): Promise<PagedResult<AuditLogDto>> {
    const query = new GetAuditLogsQuery(
      queryDto.entityName,
      queryDto.entityId,
      queryDto.userId,
      queryDto.action,
      queryDto.fromDate ? new Date(queryDto.fromDate) : undefined,
      queryDto.toDate ? new Date(queryDto.toDate) : undefined,
      queryDto.page,
      queryDto.pageSize,
      queryDto.sortBy,
      queryDto.sortDir,
      queryDto.search,
    );

    return this.queryBus.execute(query);
  }
}

File: backend/dotnet/clean-architecture/full/src/Presentation/App.Template.WebAPI/Controllers/AuditLogsController.cs
using AppTemplate.Application.Common.Models;
using AppTemplate.Application.DTOs;
using AppTemplate.Application.Features.AuditLogManagement.Queries.GetAuditLogs;
using MediatR;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace AppTemplate.WebAPI.Controllers;

/// <summary>
/// Audit log management endpoints (Admin only)
/// </summary>
[Authorize(Roles = "Admin")]
[ApiController]
[Route("api/audit-logs")]
public class AuditLogsController : ControllerBase
{
    private readonly ISender _mediator;

    public AuditLogsController(ISender mediator)
    {
        _mediator = mediator;
    }

    /// <summary>
    /// Get list of audit logs with pagination and filtering
    /// </summary>
    [HttpGet]
    [ProducesResponseType(typeof(PagedResult<AuditLogDto>), StatusCodes.Status200OK)]
    public async Task<IActionResult> GetAuditLogs(
        [FromQuery] int page = 1,
        [FromQuery] int pageSize = 20,
        [FromQuery] string? sortBy = null,
        [FromQuery] string? sortDir = "desc",
        [FromQuery] string? search = null,
        [FromQuery] string? entityName = null,
        [FromQuery] string? entityId = null,
        [FromQuery] string? userId = null,
        [FromQuery] string? action = null,
        [FromQuery] DateTime? fromDate = null,
        [FromQuery] DateTime? toDate = null)
    {
        var query = new GetAuditLogsQuery
        {
            Page = Math.Max(1, page),
            PageSize = Math.Clamp(pageSize, 1, 100),
            SortBy = sortBy,
            SortDir = sortDir,
            Search = search,
            EntityName = entityName,
            EntityId = entityId,
            UserId = userId,
            Action = action,
            FromDate = fromDate,
            ToDate = toDate
        };
        var result = await _mediator.Send(query);
        return Ok(result);
    }
}
